sql: fix zone config validation during a REGIONAL BY ROW transition

otan opened the pull request:
Involves peeping into the mutations and excluding the new indexes being
created -- these will not have the correct zone configs set on them.

Refs: #63204

Release note (bug fix): Fix a bug where
crdb_internal.validate_multi_region_zone_configs() fails during a
REGIONAL BY ROW locality transition.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63273)
<!-- Reviewable:end -->

ajstorm reviewed (changes requested):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63273)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/region_util.go, line 1336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlMhAb6qFYmaB7iWsj:-MXlMhAb6qFYmaB7iWsk:bxq0tn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1336)):*
> ```Go
> 
> 	// We must remove any REGIONAL BY ROW locality swap that is in progress,
> 	// as these zone configurations would be incorrect.
> ```

Could just be me, but all the cases we're trying to reason with here are difficult to keep straight in my head.  Can you enhance this comment to describe what scenario requires this change (we're in a RBR transition, we have indexes defined on the table, the indexes are going to have their PKs swapped, those backfilled indexes may not have the right zone configs set because we only apply them at the end, etc), and why doing what we're doing here helps?  You can kinda get this info now with a bunch of sluthing into the git history, but it would be nicer of we had it immediately available in the code.

Also, can we change `indexesToRemove` to `indexesBeingSwapped`?  Then hopefully the intent of why we're removing them will be clear throughout this function.
___
*[pkg/sql/region_util.go, line 1361 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlNxx9A1WsUUcR3vAd:-MXlNxx9A1WsUUcR3vAe:b-jq4dlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1361)):*
> ```Go
> 	}
> 
> 	// Remove inactive subzones from the comparison.
> ```

Nit: Can we update this comment?  While it's true that we're removing the inactive subzones, what the action below is really doing is obtaining a list of the active subzones.  Maybe something like "We only want to compare against the list of subzones on active indexes, so filter the subzone list based on the activeSubzoneIndexIDs computed above."?
___
*[pkg/sql/region_util.go, line 1362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlQWJ54xPikLbR7LpZ:-MXlQWJ608oj3bismp0J:b-uuiqul) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1362)):*
> ```Go
> 
> 	// Remove inactive subzones from the comparison.
> 	filteredCurrentZoneConfigSubzones := currentZoneConfig.Subzones[:0]
> ```

Nit: If you go with the comment above, can we make this `filteredActiveZoneConfigSubzones` for consistency?
___
*[pkg/sql/region_util.go, line 1375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlQzFT9qnbPjlK7cqb:-MXlQzFT9qnbPjlK7cqc:b-hd3yrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1375)):*
> ```Go
> 	}
> 
> 	// Remove irrelevant zone configs from the expectedZoneConfig.
> ```

Nit: irrelevant how?  Because they're being swapped?  If so, might want to update the comment here.  Also, having `indexesToRemove` be `indexesBeingSwapped` will help with context down here too.
___
*[pkg/sql/region_util.go, line 1386 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlRtcVBRaNRLcnEzTy:-MXlRtcVBRaNRLcnEzTz:bq6dco1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1386)):*
> ```Go
> 		}
> 	}
> 	expectedZoneConfig.Subzones = filteredExpectedZoneConfigSubzones
> ```

Nit: Would it not be clearer here to operate directly on `expectedZoneConfig.Subzones by pulling items out of the map when they're found?  Then we wouldn't need to manage this new fiteredExpectedZoneConfigSubzones.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63273)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/sql/region_util.go, line 1362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlQWJ54xPikLbR7LpZ:-MXnXsYgEojpe5HjMQ39:b-5uhb61) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1362)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: If you go with the comment above, can we make this `filteredActiveZoneConfigSubzones` for consistency?
</blockquote></details>

i think the current name is better, as it represents the current zone config.
___
*[pkg/sql/region_util.go, line 1386 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlRtcVBRaNRLcnEzTy:-MXnYF_X86V5o0_ofpgI:b-60vlw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1386)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: Would it not be clearer here to operate directly on `expectedZoneConfig.Subzones by pulling items out of the map when they're found?  Then we wouldn't need to manage this new fiteredExpectedZoneConfigSubzones.
</blockquote></details>

these is a Go-ism to re-use the same slice to filtered down elements (e.g. https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/schemachange/operation_generator.go#L1955)
https://stackoverflow.com/a/50183212


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63273)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/region_util.go, line 1336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlMhAb6qFYmaB7iWsj:-MXoUJ3f7wyEigHrso8F:b-n2le15) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1336)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Could just be me, but all the cases we're trying to reason with here are difficult to keep straight in my head.  Can you enhance this comment to describe what scenario requires this change (we're in a RBR transition, we have indexes defined on the table, the indexes are going to have their PKs swapped, those backfilled indexes may not have the right zone configs set because we only apply them at the end, etc), and why doing what we're doing here helps?  You can kinda get this info now with a bunch of sluthing into the git history, but it would be nicer of we had it immediately available in the code.

Also, can we change `indexesToRemove` to `indexesBeingSwapped`?  Then hopefully the intent of why we're removing them will be clear throughout this function.
</blockquote></details>

I like the change.  

Nit: there's an extra `a` in the first row.
___
*[pkg/sql/region_util.go, line 1362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63273#-MXlQWJ54xPikLbR7LpZ:-MXoUv-wAxywGY53jgjF:bhfna4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a64daac4f6fb33bcbdcc6406c27f631c6733754/pkg/sql/region_util.go#L1362)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i think the current name is better, as it represents the current zone config.
</blockquote></details>

With the name change above to "New", current makes more sense down here.  :thumbsup: 


<!-- Sent from Reviewable.io -->

otan committed:
sql: fix zone config validation during a REGIONAL BY ROW transition

Involves peeping into the mutations and excluding the new indexes being
created -- these will not have the correct zone configs set on them.

Release note (bug fix): Fix a bug where
crdb_internal.validate_multi_region_zone_configs() fails during a
REGIONAL BY ROW locality transition.

otan commented:
bors r=ajstorm
